### PR TITLE
Optimize Aggregator/Join/Set keys

### PR DESCRIPTION
**Pull request by solotzg**
### What problem does this PR solve?

Issue Number: ref #5294

### What is changed and how it works?

- optimize agg with one string key
  - type: str-bin, str-bin-padding
  - `struct AggregationMethodOneKeyStringNoCache`
- optimize agg with two keys
  - type: number-64-bytes, str-bin, str-bin-padding
  - `struct AggregationMethodFastPathTwoKeysNoCache`
- `AggregationMethodOneKeyStringNoCache` and `AggregationMethodFastPathTwoKeysNoCache` will generate empty data for string column because such column will be discarded.
  - https://github.com/pingcap/tiflash/blob/84c2650bc4320919b954babeceb5aeaadb845770/dbms/src/Columns/IColumn.h#L160-L173
  - https://github.com/pingcap/tiflash/blob/84c2650bc4320919b954babeceb5aeaadb845770/dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp#L272-L306
- optimize `ColumnString::deserializeAndInsertFromArena` by same way above.

----

We may need `JIT` for further optimization about AGG/JOIN/SET later.

### Benchmark

ENV

- TiFlash x 1
- limit cpu up to 2000%
- original commit: 49ca973d827175b327ca2040455b17b8ab3fe163
- [ClickBench](https://github.com/ClickHouse/ClickBench/blob/d661b49dffda015415366f0e12aa1ebe3424c311/mysql/queries.sql)

Time(s) | Original | Optimized | Improvement: (Original) / (Optimized) - 1.0
-- | -- | -- | --
ClickBench TiFlash x 1 TIFLASH-REPLICA 1
Q14 | 2.075 | 1.835 | 13.08%
Q15 | 1.21 | 1.155 | 4.76%
Q17 | 3.695 | 3.365 | 9.81%
Q18 | 2.925 | 2.805 | 4.28%
Q34 | 6.4 | 6.085 | 5.18%
Q35 | 6.575 | 6.28 | 4.70%
TPCH-100 TiFlash x 1 TIFLASH-REPLICA 1
Q1 | 7.28 | 7.08 | 2.82%

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Interpreters/Aggregator.h [hunk 1]
```diff
@@ -661,9 +806,17 @@ struct AggregatedDataVariants : private boost::noncopyable
     M(keys128)                                         \
     M(keys256)                                         \
     M(serialized)                                      \
-    M(multi_key_string)                                \
```
Comment (windtalker) [inline dbms/src/Interpreters/Aggregator.h:664]: Why delete this?

File: dbms/src/Flash/tests/gtest_aggregation_executor.cpp [hunk 2]
```diff
@@ -322,6 +342,109 @@ try
 }
 CATCH
 
+TEST_F(ExecutorAggTestRunner, AggregationCounttzg)
```
[the remark this answers is not in the document]
Comment (solotzg) [reply dbms/src/Flash/tests/gtest_aggregation_executor.cpp:345]: That's a mistake.

File: dbms/src/Interpreters/Aggregator.h (hunk 1 above)
Comment (solotzg) [reply dbms/src/Interpreters/Aggregator.h:664]: For 2 agg keys(str,str), it has been optimized by `HashMethodFastPathTwoKeys`. It is used in tpch q1.
For other cases with GE than 3 agg keys, we may need to design and optimize later.

Comment (windtalker) [reply dbms/src/Interpreters/Aggregator.h:664]: OK, so for more than 3 agg string key, it now will use `serialized`?

File: dbms/src/Common/ColumnsHashing.h [hunk 3]
```diff
@@ -223,6 +251,123 @@ struct HashMethodMultiString
...
+    {
+        return getKeyImpl(row, ref, [](StringRef key) {
+            return DB::BinCollatorSortKey<true>(key.data, key.size);
+        });
+    }
+};
+
+/// For the case when there are 2 keys.
+template <typename Key1Desc, typename Key2Desc, typename Value, typename Mapped>
+struct HashMethodFastPathTwoKeys
```
[the remark this answers is not in the document]
Comment (solotzg) [reply dbms/src/Common/ColumnsHashing.h:341]: Done.

File: dbms/src/Columns/ColumnString.h [hunk 4]
```diff
@@ -220,24 +217,34 @@ class ColumnString final : public COWPtrHelper<IColumn, ColumnString>
...
+            offsets.push_back(0);
+            return pos + string_size;
+        }
+        else
+        {
+            const size_t old_size = chars.size();
+            const size_t new_size = old_size + string_size;
+            chars.resize(new_size);
+            inline_memcpy(&chars[old_size], pos, string_size);
+            offsets.push_back(new_size);
```
[the remark this answers is not in the document]
Comment (solotzg) [reply dbms/src/Columns/ColumnString.h:245]: Done. Use insertDataWithTerminatingZero.

File: dbms/src/Interpreters/Aggregator.h (hunk 1 above)
Comment (solotzg) [reply dbms/src/Interpreters/Aggregator.h:664]: Yes.

Comment (solotzg) [reply dbms/src/Interpreters/Aggregator.h:664]: If there is any typical workload (like tpch or clickbench) use agg with 3 str keys, maybe we should add such method back.

**Review by windtalker — APPROVED** [review]
LGTM

**Review by zanmato1984 — APPROVED** [review]
LGTM

**Pull request merged**